### ISSUE TITLE
GatewayId is evil

### ISSUE THREAD

**Issue by pas256** [cloudtools/troposphere#436]
Hey there,

I just spent a bit of time with AWS Support, so wanted to pass it on and get a type check in Route.

troposphere will gladly generate a template from this code:

```python
nat = t.add_resource(ec2.NatGateway(
    'NatGateway{}'.format(letter),
    AllocationId=GetAtt(eip, 'AllocationId'),
    SubnetId=az['subnet_id']
))
t.add_resource(ec2.Route(
    'NatGateway{}Route'.format(letter),
    GatewayId=Ref(nat),
    DestinationCidrBlock='0.0.0.0/0',
    RouteTableId=Ref(route_table_id),
))
```

and silly me didn't realize that `GatewayId` is only for IGWs and VGWs. For a NAT Gateway they have a separate field called `NatGatewayId` when creating a Route. The above should be:

```
t.add_resource(ec2.Route(
    'NatGateway{}Route'.format(letter),
    NatGatewayId=Ref(nat),
    DestinationCidrBlock='0.0.0.0/0',
    RouteTableId=Ref(route_table_id),
))
```

I think some troposhere resources do some type checking, right? Can we add a type check to `GatewayId` in `Route` to return an error message when someone references a `ec2.NatGateway`?

**Comment by rcuza** [comment]
I agree, it would of been easier if they had just overloaded GatewayId with yet another use. I wonder why AWS didn't.

In addition to this code producing a CF template without an error, the `boto3.client.validate_template(TemplateBody=t.to_json())` also does not return any errors. This is a good problem for a validator because the only way to discover it is to actually build the stack and watch it rollback.

For troposphere to validate this would require "knowing" that the route resource will be used with either a NatGateway or IGW/VGW. In the troposhpere code, NatGatewayId, like most troposphere object's, attributes is just a `basestring`. Writing your solution as pseudo code:

```
for every Route resource:
    if it has a GatewayId:
        if the GatewayId setting is a NatGateway:
            return error
```

This seems easy if `GatewayId=<troposphere NatGateway obj>`, but if it is just a `string`, then I'm not sure how this can be discovered. What if the script is referencing a resource that already exists? Maybe the validation can only check when the attribute is set to an object.

**Comment by rcuza** [comment]
Just took a quick stab at it and I don't think `validators.py` will work. If one passes a `<troposphere.ec2.NatGateway>` object, then it would be straight forward. But in your example, and in how most people build troposphere scripts, you passed a `<troposphere.Ref>` object. This only returns the string name of the object being referenced (as I understand it). The validator would have to have a higher perspective on the script being written to be able to trace the string to the object. 

I am not sure troposphere is the right place to solve this problem, at least with its current features. If you have another idea for solving this problem, that would be helpful.

---
Here are the changes I made while thinking about the problem:
 * Change 'GatewayId' in `ec2.py`:
```
class Route(AWSObject):
    resource_type = "AWS::EC2::Route"

    props = {
        'DestinationCidrBlock': (basestring, True),
        'GatewayId': (gatewayobjstring, False),
        'InstanceId': (basestring, False),
        'NatGatewayId': (basestring, False),
        'NetworkInterfaceId': (basestring, False),
        'RouteTableId': (basestring, True),
        'VpcPeeringConnectionId': (basestring, False),
    }
```
* Add `gatewayobjstring` to `validators.py`:
```
def gatewayobjstring(x):
    """
    Return a string or obj if x isn't a NatGateway Obj
    """
    if type(x) is str:
        return x
    elif type(x) is troposphere.ec2.NatGateway:
        raise ValueError("%r is not valid. NatGW is not IGW or VGW" % x)
    elif type(x) is troposphere.Ref:
        if we_could_only_tell_what_the_original_object_was:
    else:
        return x
```
